File: src/comm_setup.py
"""
Algorithm 2: Communication Information Collection
==================================================

Paper reference: Section III.C, Algorithm 2.

Given the diagonal block boundaries from Algorithm 1, this module builds
the communication schedule: which vector elements each process must send
to / receive from every other process.

Key data structures built here:

  sendid[p] : np.ndarray[int32]
      Column indices that *this* process must send to process p.
      These are columns we own (our initial row range) that p needs.

  recvid[p] : np.ndarray[int32]
      Column indices that *this* process must receive from process p.
      These are columns p owns that appear in our off-diagonal block.

  recv_idx_map[p] : Dict[int, int]
      Maps global column → local position in the per-process recv buffer.
      Used during unpacking (Algorithm 3).

Column ownership is determined by the **initial row distribution**:
column c is owned by the process whose r_start ≤ c < r_end.  This is
deterministic and invariant under boundary expansion.

Communication pattern (implemented inside ``build_schedule``):

  1. MPI_Allgather — share all processes' (left, right, r_start, r_end).
  2. Local scan — build recvid per remote process, deduplicating columns.
  3. MPI_Alltoall — exchange recv *counts* (how many columns we need from
     each peer).  This tells each process how many columns it must *send*.
  4. MPI_Isend / MPI_Irecv — exchange the actual column-index lists.
  5. Reconstruct sendid from received lists.
"""

from __future__ import annotations

import logging
from typing import Dict, List, Tuple

import numpy as np

logger = logging.getLogger(__name__)


# ──────────────────────────────────────────────────────────────────────
# Public API
# ──────────────────────────────────────────────────────────────────────

def build_schedule(
    rowptr: np.ndarray,        # int32, shape (nlocal+1,)
    colidx: np.ndarray,        # int32, shape (nnz_local,)
    r_start: int,
    r_end: int,
    left: int,
    right: int,
    comm,
) -> Tuple[Dict[int, np.ndarray], Dict[int, np.ndarray], np.ndarray]:
    """Build communication schedule (Algorithm 2).

    Parameters
    ----------
    rowptr, colidx : np.ndarray
        CSR data for locally-owned rows.
    r_start, r_end : int
        This process's initial global row range (determines column ownership).
    left, right : int
        Diagonal block boundaries from Algorithm 1.
    comm :
        MPI communicator.

    Returns
    -------
    sendid : Dict[int, np.ndarray]
        sendid[q] = columns this rank must send to rank q (int32).
    recvid : Dict[int, np.ndarray]
        recvid[q] = columns this rank must receive from rank q (int32).
    recv_idx_map : np.ndarray of dict
        recv_idx_map[q] maps global col → local position in recv buffer for q.
    """
    rank = comm.Get_rank()
    nprocs = comm.Get_size()
    nlocal = r_end - r_start

    # ── Step 1: Allgather boundaries & initial row ranges ──
    my_info = np.array([left, right, r_start, r_end], dtype=np.int64)
    all_info = np.empty((nprocs, 4), dtype=np.int64)
    comm.Allgather(my_info, all_info)

    all_left = all_info[:, 0]
    all_right = all_info[:, 1]
    all_r_start = all_info[:, 2]
    all_r_end = all_info[:, 3]

    # ── Step 2: Build recvid (what we need from each remote process) ──
    # recv_sets[q] = set of global columns we need from process q
    recv_sets: Dict[int, set] = {}
    for q in range(nprocs):
        if q != rank:
            recv_sets[q] = set()

    for i in range(nlocal):
        for j in range(rowptr[i], rowptr[i + 1]):
            c = int(colidx[j])
            if left <= c < right:
                continue  # diagonal — locally available, no comm needed
            owner = _find_owner(c, all_r_start, all_r_end)
            if owner == rank:
                # We own this column even though it's outside our diagonal
                # block — this can happen due to boundary expansion.
                continue
            if owner not in recv_sets:
                recv_sets[owner] = set()
            recv_sets[owner].add(c)

    # Convert sets to sorted int32 arrays; build index map
    recvid: Dict[int, np.ndarray] = {}
    recv_idx_map: Dict[int, Dict[int, int]] = {}
    for q, cols in recv_sets.items():
        if len(cols) == 0:
            continue
        sorted_cols = np.array(sorted(cols), dtype=np.int32)
        recvid[q] = sorted_cols
        # Map col → position in this per-peer buffer
        recv_idx_map[q] = {int(c): idx for idx, c in enumerate(sorted_cols)}

    # ── Step 3: Alltoall to exchange counts ──
    # send_counts[q] = |recvid[q]| (how many columns we need from q)
    send_counts = np.zeros(nprocs, dtype=np.int32)
    for q, cols in recvid.items():
        send_counts[q] = len(cols)
    recv_counts = np.zeros(nprocs, dtype=np.int32)
    comm.Alltoall(send_counts, recv_counts)
    # recv_counts[q] = how many columns process q needs from US
    # (i.e. |sendid[q]| we need to build)

    # ── Step 4: Exchange column-index lists via point-to-point ──
    # We send recvid[q] to process q; we receive sendid[q] from process q.
    sendid: Dict[int, np.ndarray] = {}

    from mpi4py import MPI

    requests: list = []

    # Post receives for incoming column lists (what others need from us)
    recv_buffers: Dict[int, np.ndarray] = {}
    for q in range(nprocs):
        if q == rank:
            continue
        cnt = int(recv_counts[q])
        if cnt > 0:
            buf = np.empty(cnt, dtype=np.int32)
            recv_buffers[q] = buf
            req = comm.Irecv(buf, source=q, tag=100 + q)
            requests.append(req)

    # Post sends of our recvid lists to each peer
    for q, cols in recvid.items():
        req = comm.Isend(cols, dest=q, tag=100 + rank)
        requests.append(req)

    # Wait for all exchanges to complete
    MPI.Request.Waitall(requests)

    # Convert received buffers into sendid
    for q, buf in recv_buffers.items():
        sendid[q] = buf

    # ── Statistics ──
    total_recv = sum(len(v) for v in recvid.values())
    total_send = sum(len(v) for v in sendid.values())
    logger.info(
        "Algorithm 2: send to %d peers (%d elems), recv from %d peers (%d elems)",
        len(sendid), total_send, len(recvid), total_recv,
    )

    return sendid, recvid, recv_idx_map


# ──────────────────────────────────────────────────────────────────────
# Helpers
# ──────────────────────────────────────────────────────────────────────

def _find_owner(col: int, all_r_start: np.ndarray, all_r_end: np.ndarray) -> int:
    """Return the rank that owns global column *col*.

    Ownership follows the initial row distribution: process q owns
    columns [all_r_start[q], all_r_end[q]).
    """
    for q in range(len(all_r_start)):
        if all_r_start[q] <= col < all_r_end[q]:
            return q
    # Fallback: column outside any range (shouldn't happen for square matrices)
    # Assign to last process
    return len(all_r_start) - 1

File: src/comm_setup.cpp
#include "comm_setup.hpp"

#include <algorithm>
#include <cstdio>
#include <cstring>
#include <set>
#include <unordered_map>
#include <vector>

namespace distspmv {

// Helper: find owner rank for global column c (binary search on monotonic ranges)
static int find_owner(idx_t c, const idx_t* all_r_start,
                      const idx_t* all_r_end, int nprocs) {
    auto it = std::upper_bound(all_r_start, all_r_start + nprocs, c);
    if (it == all_r_start) return 0;
    int q = static_cast<int>(it - all_r_start) - 1;
    return (c < all_r_end[q]) ? q : nprocs - 1;
}

CommSchedule build_schedule(
    const idx_t* rowptr, const idx_t* colidx,
    idx_t nlocal, idx_t r_start, idx_t r_end,
    idx_t left, idx_t right, MPI_Comm comm) {

    int rank, nprocs;
    MPI_Comm_rank(comm, &rank);
    MPI_Comm_size(comm, &nprocs);

    // ── Step 1: Allgather boundaries & initial row ranges ──
    idx_t my_info[4] = {left, right, r_start, r_end};
    std::vector<idx_t> all_info(static_cast<std::size_t>(nprocs) * 4);
    MPI_Allgather(my_info, 4, MPI_INT, all_info.data(), 4, MPI_INT,
                  comm);

    std::vector<idx_t> all_left(nprocs), all_right(nprocs);
    std::vector<idx_t> all_r_start(nprocs), all_r_end(nprocs);
    for (int q = 0; q < nprocs; ++q) {
        all_left[q]   = all_info[q * 4 + 0];
        all_right[q]  = all_info[q * 4 + 1];
        all_r_start[q] = all_info[q * 4 + 2];
        all_r_end[q]  = all_info[q * 4 + 3];
    }

    // ── Step 2: Build recvid sets (columns needed from each remote rank) ──
    std::unordered_map<int, std::set<idx_t>> recv_sets;
    for (int q = 0; q < nprocs; ++q) {
        if (q != rank) recv_sets[q] = std::set<idx_t>();
    }

    for (idx_t i = 0; i < nlocal; ++i) {
        for (idx_t j = rowptr[i]; j < rowptr[i + 1]; ++j) {
            idx_t c = colidx[j];
            if (left <= c && c < right) continue;  // diagonal block

            int owner = find_owner(c, all_r_start.data(), all_r_end.data(),
                                   nprocs);
            if (owner == rank) continue;  // we own it

            recv_sets[owner].insert(c);
        }
    }

    // Build recvid arrays and recv_idx_map
    CommSchedule sched;

    for (auto& [q, cols] : recv_sets) {
        if (cols.empty()) continue;
        sched.recvid[q].assign(cols.begin(), cols.end());
        // recvid is already sorted (set is ordered)
        for (std::size_t idx = 0; idx < sched.recvid[q].size(); ++idx) {
            sched.recv_idx_map[q][sched.recvid[q][idx]] = static_cast<int>(idx);
        }
    }

    // ── Step 3: Alltoall exchange counts ──
    std::vector<int> send_counts(nprocs, 0);
    for (auto& [q, cols] : sched.recvid) {
        send_counts[q] = static_cast<int>(cols.size());
    }
    std::vector<int> recv_counts(nprocs, 0);
    MPI_Alltoall(send_counts.data(), 1, MPI_INT, recv_counts.data(), 1, MPI_INT,
                 comm);
    // recv_counts[q] = how many columns rank q needs from US (|sendid[q]|)

    // ── Step 4: Exchange column-index lists via non-blocking P2P ──
    std::vector<MPI_Request> requests;
    std::unordered_map<int, std::vector<idx_t>> recv_bufs;

    // Post receives for incoming column lists
    for (int q = 0; q < nprocs; ++q) {
        if (q == rank) continue;
        int cnt = recv_counts[q];
        if (cnt > 0) {
            recv_bufs[q].resize(cnt);
            MPI_Request req;
            MPI_Irecv(recv_bufs[q].data(), cnt, MPI_INT, q, 100 + q,
                      comm, &req);
            requests.push_back(req);
        }
    }

    // Post sends of our recvid lists
    for (auto& [q, cols] : sched.recvid) {
        MPI_Request req;
        MPI_Isend(cols.data(), static_cast<int>(cols.size()), MPI_INT,
                  q, 100 + rank, comm, &req);
        requests.push_back(req);
    }

    // Wait for all exchanges
    MPI_Waitall(static_cast<int>(requests.size()), requests.data(),
                MPI_STATUSES_IGNORE);

    // Build sendid from received buffers
    for (auto& [q, buf] : recv_bufs) {
        if (!buf.empty()) {
            sched.sendid[q] = std::move(buf);
        }
    }

    // ── Statistics ──
    int64_t total_recv = 0, total_send = 0;
    for (auto& [_, cols] : sched.recvid) total_recv += cols.size();
    for (auto& [_, cols] : sched.sendid) total_send += cols.size();
    if (rank == 0) {
        std::printf("[rank %2d] Algo2: send to %zu peers (%lld elems), "
                    "recv from %zu peers (%lld elems)\n",
                    rank, sched.sendid.size(), total_send,
                    sched.recvid.size(), total_recv);
    }

    return sched;
}

}  // namespace distspmv
